### ISSUE TITLE
Indicate language server progress

### ISSUE THREAD

**Issue by ul** [ul/kak-lsp#38]
Some of language servers take a long time to initialize. Most of them report progress and/or ready event. We need to indicate their progress somehow to prevent confusion like "why does 'go to definition' not work? is it broken?" during initialization.

I need help to understand what is the best way in Kakoune to communicate such kind of information. It should be non-intrusive and discoverable at the same time. Should it be `echo` or `info` (the former could be easily missed and the latter feels too noisy)? Ideal would be to show in status line (like `echo`) but persistent and update according to language server status change.

**Comment by occivink** [comment]
I think exposing a option for the user to put in their modeline is a good solution. It could be empty when idle and contain a simple indicator when something is in progress.

**Comment by ervasive** [comment]
Another option could be to show an infobox when the user calls 'go to definition' command so it only informs in case of action.

**Comment by ul** [comment]
Thanks, both tactics should work well together: option with current status for those who wants to put it into modeline or do any other custom notification and info which explains why actions couldn't be performed right now. For the latter introducing generic error reporting mechanism could be enough (need to double check what exactly language server responds when it's not ready, because it could be not an error but just empty list of declarations/references/etc. In that case we need to track ready event in kak-lsp itself and do smart guess).

**Comment by Delapouite** [comment]
I tried to use `:lsp-references` with no apparent success on a JavaScript project.
It turns out that:

```
May 30 12:20:52.463 ERRO Error response from server: Failure { jsonrpc: Some(V2), error: Error { code: MethodNotFound, message: "Unhandled method textDocument/references", data: None }, id: Num(4) }, module: kak_lsp::controller:212
```

So the problem is that [flow-language-server](https://github.com/flowtype/flow-language-server) does not support `textDocument/references` (yet).

No big deal, the feature will come in due time.

But the lack of error feedback on kakoune side was a bit disturbing.

Here are a few suggestions on top of the great ones in this thread:
- add a line with something like `kak-lsp: no support for lsp-references with flow-language-server` in the `*debug*` buffer.
- add a new command like `lsp-capabilities` to show what stuffs the current server support.

**Comment by ul** [comment]
I've added reporting server errors (including unsupported methods) back to editor, both writing them to `*debug*` buffer and showing in info box (7ec2e8a9dc92cf8d0ad3d740bee0a53c95e185d8). For some of language servers (like cquery) it solves problem of indicating progress because they report an error if they are not ready to give proper response. For others (like rls) it doesn't, because when they are not ready they just give wrong empty response. Progress reporting itself is highly inconsistent across language servers and sometimes have no clear signal of when language server is ready.

I created #51 for `lsp-capabilities`.

I'm going to close this issue because problem is mostly solved for "good citizen" type of language servers, for others it should be solved on case-by-case basis. Feel free submit either issue with description how you see progress reporting for language server you use should look like or PR implementing that behaviour. Of course, any ideas about how to improve current "generic" error reporting are welcome as well.
